Interleave downloads with installations when downloading a toolchain

FranciscoTGouveia opened the pull request:
Continuation of #4436.

This change aims to speed up the overall performance of the `rustup [toolchain] install` command by allowing the installation of a component to start immediately after its download completes.
This approach is particularly beneficial for slower connections, as it enables installations to progress in parallel with ongoing downloads -- so by the time all downloads are finished, all installations should be completed as well.

**NB:** This PR is not yet ready for review. I believe this change should be backed by benchmarks and user feedback before moving forward. I’ll share those here in the next couple of days.

FranciscoTGouveia committed:
feat(downloads): interleave the downloads with their installations

Even though downloads are done concurrently, the installations are done
sequentially. This means that, as downloads complete, they are in a
queue (an mpsc channel) waiting to be consumed by the future responsible
for the (sequential) installations.

There was a need to relax some test cases to allow for uninstall to
happen before the downloads.

rami3l commented on src/dist/manifestation.rs:
```diff
@@ -246,17 +213,94 @@ impl Manifestation {
+                        if counter >= total_components {
+                            break;
+                        }
+                        if let Some(message) = download_rx.recv().await {
```
Could it be that the counter never gets incremented because `.recv().await` returns `None`?

Did you mean something like this instead?

```rs
while counter < total_components && let Some(_) = ... {  ...; counter += 1; }
```

rami3l commented on src/dist/manifestation.rs:
```diff
@@ -246,17 +213,94 @@ impl Manifestation {
+                                        Ok(new_tx) => {
+                                            current_tx = new_tx;
+                                        }
+                                        Err(e) => {
```
If I'm not mistaken, `Err(e) => return Err(e)` is just `?` 🤔

rami3l commented on tests/suite/cli_rustup.rs:
```diff
@@ -35,7 +35,6 @@ async fn rustup_stable() {
         .with_stderr(snapbox::str![[r#"
 info: syncing channel updates for 'stable-[HOST_TRIPLE]'
 info: latest update on 2015-01-02, rust version 1.1.0 (hash-stable-1.1.0)
-info: downloading component[..]
```
Possibly squash this commit with the one that has broken these tests.

rami3l commented on src/dist/notifications.rs:
```diff
@@ -122,6 +124,13 @@ impl Display for Notification<'_> {
                 }
             }
+            ComponentInstalled(c, h, t) => {
+                if Some(h) == t.as_ref() || t.is_none() {
```
Nit: Now that we are at it, I would probably give an extra commit to avoid the `t.unwrap()` here by flipping the `if` and writing something like:

```rs
if let Some(t) = t && t != h { ... } else { ... }
```

FranciscoTGouveia replied:
I did this based on how other Notifications were being handled.
Should I refactor all of them now, leave that for another PR, or just change this one?

rami3l replied:
@FranciscoTGouveia It should be worthwhile to refactor them all in a separate commit (before your addition) lest we forget about it :)

FranciscoTGouveia replied:
If you agree, I would like to leave this for a follow-up PR, right after this one is merged.
So I suggest leaving this as unresolved for now.

rami3l replied:
Okay, let's do https://github.com/rust-lang/rustup/pull/4471#discussion_r2362667381 for the particular change in this PR, and then move the rest to that style in a subsequent PR.

rami3l commented on src/cli/download_tracker.rs:
```diff
@@ -170,8 +174,11 @@ impl DownloadTracker {
+                ProgressStyle::with_template( if pb.position() != 0 {
+                    "{msg:>12.bold}  downloaded {total_bytes} in {elapsed} installing now {spinner:.green}"
+                } else {
+                    "{msg:>12.bold}  component already downloaded, installing now {spinner:.green}"
```
Nit: the `, installing now`, ` installing now` and ` and installed` might seem inconsistent. How about just ` and installed` and ` and installing`? I believe the spinner can compensate for the minor difference.

rami3l commented on src/dist/manifestation.rs:
```diff
@@ -224,79 +224,104 @@ impl Manifestation {
+                                    component,
+                                    format,
+                                    installer_file,
+                                    Arc::clone(&tmp_cx),
```
Note: Looks like there are extra clones already... So in addition to making sure the new implementation is working correctly, it would also be useful to audit the use of `Arc::clone()`.

FranciscoTGouveia replied:
Thanks for noticing this!
I will revisit all uses of `Arc::clone()` now.

FranciscoTGouveia committed:
feat(downloads): use `spawn_blocking` to allow the downloads to continue during installations

Co-authored-by: rami3l <rami3l@outlook.com>

FranciscoTGouveia commented:
As noted in the opening message of the PR, I now present some benchmarks that will enable us to move forward with this change.
As stated before, this will bump the install and update commands' performance, as when the downloads finish, the installations have already executed.

The results are presented in the plot below, for both low-speed and high-speed internet connections.

<img width="1896" height="986" alt="image" src="https://github.com/user-attachments/assets/bf876bdf-ab32-43ab-a36b-3a1bbdd6c2f6" />

These results were obtained by running the `rustup toolchain install nightly` command 10x times for each combination of variables (`RUSTUP_CONCURRENT_DOWNLOADS`, interleaved or not, and internet connection).
Please note that the *y axis* do not start on zero, and cannot be directly compared between themselves.

---
Below, I also leave a small animation of the previous and current behavior of the `rustup toolchain install` command to illustrate the *UX* changes that this PR also implies.
On the left you can see the previous version and on the right, the version with the interleaved installations.
Please note that this is just an animation and does not have the purpose of showing any kind of speedup.

![animation4471-ezgif com-speed](https://github.com/user-attachments/assets/d3b44b2c-afd6-4ab1-8edf-17d0eabcd0ba)

Thanks for taking the time to review this PR!

rami3l commented on src/config.rs:
```diff
@@ -330,13 +330,13 @@ impl<'a> Cfg<'a> {
         DownloadCfg {
             dist_root: &self.dist_root_url,
-            tmp_cx: &self.tmp_cx,
+            tmp_cx: Arc::new(self.tmp_cx.clone()),
```
What if we make `self.tmp_cx` an `Arc<>` directly?

rami3l commented on src/config.rs:
```diff
@@ -330,13 +330,13 @@ impl<'a> Cfg<'a> {
+            tmp_cx: Arc::new(self.tmp_cx.clone()),
             download_dir: &self.download_dir,
-            notify_handler,
+            notify_handler: Arc::clone(&notify_handler),
```
This `Arc::clone()` is redundant.

rami3l commented on src/dist/component/package.rs:
```diff
@@ -551,36 +552,36 @@ impl Package for TarPackage<'_> {
         Ok(TarGzPackage(TarPackage::new(
             stream,
-            tmp_cx,
+            Arc::clone(&tmp_cx),
```
This `Arc::clone()` is redundant.

rami3l commented on src/dist/component/package.rs:
```diff
@@ -589,36 +590,36 @@ impl Package for TarGzPackage<'_> {
         Ok(TarXzPackage(TarPackage::new(
             stream,
-            tmp_cx,
+            Arc::clone(&tmp_cx),
```
This `Arc::clone()` is redundant.

rami3l commented on src/dist/component/package.rs:
```diff
@@ -627,36 +628,36 @@ impl Package for TarXzPackage<'_> {
         Ok(TarZStdPackage(TarPackage::new(
             stream,
-            tmp_cx,
+            Arc::clone(&tmp_cx),
```
This `Arc::clone()` is redundant.

rami3l commented on src/dist/download.rs:
```diff
@@ -16,15 +17,48 @@ use crate::utils;
 }
 
+#[derive(Clone)]
+pub struct OwnedDownloadCfg {
```
Consider removing `OwnedDownloadCfg` and make `DownloadCfg` use `Arc<str>`, `Arc<Path>` and `Arc<Process>` instead, so the lifetime parameters can be removed.

rami3l commented on src/dist/temp.rs:
```diff
@@ -185,7 +191,7 @@ impl Context {
                     .with_context(|| CreatingError::File(PathBuf::from(&temp_file)))?;
                 return Ok(File {
-                    cfg: self,
+                    cfg: Arc::clone(&self),
```
This `Arc::clone()` is redundant.

rami3l commented on src/dist/temp.rs:
```diff
@@ -159,18 +165,18 @@ impl Context {
                     .with_context(|| CreatingError::Directory(PathBuf::from(&temp_dir)))?;
                 return Ok(Dir {
-                    cfg: self,
+                    cfg: Arc::clone(&self),
```
This `Arc::clone()` is redundant.

rami3l commented on src/dist/temp.rs:
```diff
@@ -1,8 +1,11 @@
 use std::io;
 use std::ops;
 use std::path::{Path, PathBuf};
+use std::{
```
Consider splitting the import here for consistency.

rami3l commented on src/dist/manifestation.rs:
```diff
@@ -224,79 +224,101 @@ impl Manifestation {
+                        let new_tx = {
+                            let this = Arc::clone(&self);
+                            let new_manifest = Arc::clone(&new_manifest);
+                            let tmp_cx = download_cfg.tmp_cx.clone();
```
Nit: It is preferable to use explicit `Arc::clone()` for such variables.

rami3l commented on src/cli/common.rs:
```diff
@@ -126,14 +125,14 @@ pub(crate) fn read_line(process: &Process) -> Result<String> {
 pub(super) struct Notifier {
     tracker: Mutex<DownloadTracker>,
-    ram_notice_shown: RefCell<bool>,
+    ram_notice_shown: RwLock<bool>,
```
Given the sheer upcoming size of that one particular commit, it might be better to split all changes regarding the replacement of RefCell to another commit, just to make the review process easier.

rami3l reviewed (commented):
Modulo the previously requested changes and the two extra nits, I think this PR is quite ready. Thanks a lot, @FranciscoTGouveia!

FranciscoTGouveia committed:
refactor(downloads): substitute RefCells with RwLocks for thread safety

FranciscoTGouveia committed:
refactor(downloads): remove some lifetime parameters to allow multi-thread installations

Co-authored-by: rami3l <rami3l@outlook.com>

djc commented on src/dist/manifestation.rs:
```diff
@@ -591,6 +541,77 @@ impl Manifestation {
+        tmp_cx: &temp::Context,
+        download_cfg: &DownloadCfg<'_>,
+        new_manifest: &Manifest,
+        tx: Transaction<'a>,
```
I think the number of arguments is here too large, and this should be rethought to have a named field `struct` that can hold all/most of the arguments.

I think the pre-existing code here could benefit from some refactoring as well, for example by packaging together the arguments passed into the compression wrappers. The uninitialized variables structure also sounds out as pretty unidiomatic.

FranciscoTGouveia replied:
Thanks for the comment and for the associated PR to simplify this!
I believe that that PR addressed everything; also, *clippy* is no longer complaining.
If you agree, we can resolve this conversation.

djc commented on src/dist/manifestation.rs:
```diff
@@ -248,17 +215,76 @@ impl Manifestation {
+                    let download_tx_cloned = download_tx.clone();
+                    async move {
+                        let _permit = sem.acquire().await.unwrap();
+                        self.download_component(
```
Pre-existing, but this argument list is also just too long, and I think we should find a different way to express this before layering on more complexity.

FranciscoTGouveia replied:
Similar to the previous comment, I believe PR 4489, which you’ve already merged, takes care of this.
If you agree, we can resolve this conversation.

djc commented on src/dist/manifestation.rs:
```diff
@@ -248,17 +215,76 @@ impl Manifestation {
+
+            let semaphore = Arc::new(Semaphore::new(concurrent_downloads));
+            let component_stream =
+                tokio_stream::iter(components.into_iter()).map(|(component, format, url, hash)| {
```
Suggest writing this closure as a named function.

rami3l replied:
@djc I think `Self::foo()` that @FranciscoTGouveia came up might a bit surprising here. Maybe you meant using regular function-in-function declarations?

Idem for https://github.com/rust-lang/rustup/pull/4471#discussion_r2362662131 and https://github.com/rust-lang/rustup/pull/4471#discussion_r2362662465.

djc commented on src/dist/manifestation.rs:
```diff
@@ -248,17 +215,76 @@ impl Manifestation {
+
+            let mut stream = component_stream.buffered(components_len);
+            let download_handle = async {
+                let mut hashes = Vec::new();
```
Suggest writing this as a named function.

djc replied:
I really think the named functions should be introduced with this commit, rather than at the end.

djc commented on src/dist/manifestation.rs:
```diff
@@ -248,17 +215,76 @@ impl Manifestation {
+                }
+                hashes
+            };
+            let install_handle = async {
```
Suggest writing this as a named function.

djc commented on src/cli/download_tracker.rs:
```diff
@@ -151,4 +155,24 @@ impl DownloadTracker {
+            .keys()
+            .find(|comp| comp.contains(component))
+            .cloned();
+        if let Some(key) = key
```
Nit: invert these for early returns.

FranciscoTGouveia replied:
I believe the recent PRs involving the simplification of the notification system already addressed this.
 If you agree, we can resolve this conversation.

djc replied:
Sure, feel free to resolve any of my comments that you think are no longer applicable!

djc commented on src/cli/download_tracker.rs:
```diff
@@ -168,7 +171,28 @@ impl DownloadTracker {
+    }
+
+    /// Notifies self that the component has been installed.
+    pub(crate) fn component_installed(&mut self, component: &str) {
```
Can we share the duplicate code?

djc commented on src/dist/notifications.rs:
```diff
@@ -122,6 +124,15 @@ impl Display for Notification<'_> {
+                    write!(f, "installing component '{c}' for '{t}'")
+                } else {
+                    write!(f, "installing component '{c}'")
+                }
```
Nit: I think this would look better as a `match`:

```rust
match t {
    Some(t) if != h => write!(f, ..),
    _ => write!(f, ..),
}
```

djc commented on src/cli/download_tracker.rs:
```diff
@@ -133,8 +133,12 @@ impl DownloadTracker {
+            } else {
+                "{msg:>12.bold}  component already downloaded"
+            })
+            .unwrap(),
```
Again there's a lot of duplication here, suggest adding a little abstraction to make it more obvious what happens here.

FranciscoTGouveia replied:
As discussed with @rami3l, I plan to redesign the progress bar styling so that they become column-aligned.

Additionally, I believe it would be useful to introduce another status update to inform the user when a download has completed and is waiting to be installed, rather than simply displaying a completed progress bar, which can be confusing.

With that in mind, this upcoming redesign will affect the highlighted code in this comment, making any changes at this stage redundant.
Since the redesign should be carefully discussed, I suggest leaving this for a future PR.

djc replied:
Sure.

djc commented on src/dist/component/transaction.rs:
```diff
@@ -34,21 +34,21 @@ use crate::utils;
-    tmp_cx: &'a temp::Context,
-    notify_handler: &'a dyn Fn(Notification<'_>),
+    changes: Vec<ChangedItem>,
+    tmp_cx: Arc<temp::Context>,
```
Seems to me that `tmp_cx`, `notify_handler` and `process` should be bundled up together.

djc commented:
Very excited about these performance improvements! Personally I don't really feel comfortable with the added complexity in the current changes -- I think there should be some more refactoring before this can be cleanly added.

rbtcollins commented:
> These results were obtained by running the rustup toolchain install nightly command 10x times for each combination of variables (RUSTUP_CONCURRENT_DOWNLOADS, interleaved or not, and internet connection).
Please note that the y axis do not start on zero, and cannot be directly compared between themselves.

As I read these results:
low speed connections : 
interleaved gives 65 seconds to 59 seconds or a 10% reduction, somewhat stable for 2 or more concurrent downloads.

high speed connections:
interleaved gives 18-12 for low concurrency and 15-13 for higher concurrency, or 33% reduction for 1 or 2 concurrent downloads, and a 13% reduction for higher concurrency, with best performance at 2 concurrent downloads, and performance decreasing for successively more downloads with interleaving, but increasing for more concurrent downloads w/out interleaving.


I suspect some interaction with memory limits showing up in the unpack phase there; I suggest raising the working memory much higher, if you're on a machine that has capacity, to see if thats the case.

But also it suggests the default should probably be quite conservative, as we still need to support unpacking on Rasberry PI class machines.

djc commented on src/dist/manifestation.rs:
```diff
@@ -249,16 +216,77 @@ impl Manifestation {
-                tx,
-            )?;
+        if components_len > 0 {
+            // Create a channel to communicate whenever a download is done and the component can be installed
+            // The `mpsc` channel was used as we need to send many messages from one producer (download's thread) to one consumer (install's thread)
+            // This is recommended in the official docs: https://docs.rs/tokio/latest/tokio/sync/index.html#mpsc-channel
```
I can't find it now, but I thought I wrote a comment on wondering why we're using a channel here, when we could instead compose the futures that download and install together.

rami3l replied:
Hmmm would you mind clarifying a little more what you meant by composing the futures?

The current constraints as we previously agreed was that the downloads are concurrent but the installations happening alongside them aren't concurrent within themselves, so I think a channel would be ideal for that case in favor of, say, a mutex.

djc replied:
So you're saying, we only do one installation at a time? If that's the case, I would suggest something more like a `FuturesUnordered` containing all the downloads, and then pulling from that to run the installations.

rami3l replied:
@djc I discussed with @FranciscoTGouveia and we would like to know: how about leaving this simplification to another PR now that we already have something working and ready to be test-driven? Leaving this outside of mainline for long would definitely mean fewer chances to test the latest changes...

djc replied:
Yeah, I can live with that.

FranciscoTGouveia commented:
Hello everyone!

I think all review comments should be addressed by now with [this](https://github.com/rust-lang/rustup/pull/4471#discussion_r2385119158) and [this](https://github.com/rust-lang/rustup/pull/4471#discussion_r2362669804) comments being left for a future PR -- the justification can be seen in the comments.

> I suspect some interaction with memory limits showing up in the unpack phase there; I suggest raising the working memory much higher, if you're on a machine that has capacity, to see if thats the case.
> 
> But also it suggests the default should probably be quite conservative, as we still need to support unpacking on Rasberry PI class machines.

First of all, thank you for your comment @rbtcollins. I have indeed reran the benchmarks with a higher memory and it didn't show any difference. For reference, it was a machine equipped with an Intel Xeon E5-2630 v2 processor and 64 GB of RAM.

Moreover, I would like to thank @djc for all the simplification work that he did, which made this PR much easier to implement. :smile:

rami3l commented on src/config.rs:
```diff
@@ -238,13 +239,13 @@ pub(crate) struct Cfg<'a> {
     pub quiet: bool,
     pub current_dir: PathBuf,
-    pub process: &'a Process,
+    pub process: &'a Process, // TODO: this may need to be inside an Arc.
```
Is this TODO finally addressed? Idem for other TODO comments in the diff.

FranciscoTGouveia replied:
I am very sorry to have left this -- it was just some notes to self during the rebase. I apologize.

rami3l commented on src/dist/component/transaction.rs:
```diff
@@ -9,8 +9,8 @@
 
-use std::fs::File;
 use std::path::{Path, PathBuf};
+use std::{fs::File, sync::Arc};
```
Nit: Maybe just merge these imports into one?

rami3l commented on src/dist/manifestation.rs:
```diff
@@ -499,6 +474,131 @@ impl Manifestation {
+                let sem = semaphore.clone();
+                let download_tx = download_tx.clone();
+                let new_manifest = Arc::clone(&new_manifest);
+                let download_cfg = download_cfg.clone();
```
Nit: We might as well use `Arc::clone()` for `download_cfg`/`dl_cfg` for the sake of consistency. This happens in a few other places as well IIRC. Suggest checking for the `remove lifetimes` commit in particular.

djc replied:
While `Arc::clone()` was popular a while ago as a method of demonstrating that it's just a cheap clone, I think much of the ecosystem has moved away from it in favor of just using `.clone()`, which is more concise.

rami3l replied:
@djc I somehow wasn't well informed then, my apologies. That should be a trivial replacement for @FranciscoTGouveia right?

FranciscoTGouveia replied:
Is this change strictly related to the highlighted lines of code, or should I go ahead and update all occurrences?

djc replied:
Change the new uses in these commits. If you want to do a sweep across the codebase that would also be nice IMO but should be in a separate PR.

FranciscoTGouveia replied:
Noted. I will make sure that this PR does not introduce any `Arc::clone()`, and I will create another PR to sweep the database soon.

FranciscoTGouveia replied:
I have checked the codebase for other instances of `Arc::clone()`, but there were none.

rami3l commented on src/dist/manifestation.rs:
```diff
@@ -419,7 +390,7 @@ impl Manifestation {
         new_manifest: &[String],
         update_hash: Option<&Path>,
-        dl_cfg: &DownloadCfg<'_>,
+        dl_cfg: Arc<DownloadCfg>, // TODO: might need to be wrapped in an Arc.
```
Nit: Redundant TODO.

rami3l reviewed (approved):
LGTM modulo some questions/nits :)

djc commented on src/dist/manifestation.rs:
```diff
@@ -499,6 +472,50 @@ impl Manifestation {
         Ok(tx)
     }
+
+    fn install_component<'a>(
```
This seems more strongly attached to `ComponentBinary` than to `Manifestation`, so would suggest attaching it to that type. Suggest ordering arguments from most specific to least specific, so `tx` probably comes after `installer_file` and `new_manifest` probably comes before `download_cfg`.

djc reviewed (commented):
So while I did say I was fine with doing spawning/channel approach, after looking at the "remove some lifetime parameters" commit I am no longer sure if that's the best approach.

Effectively it means that whoever goes and puts in place the right architecture will basically have to undo all of those changes which IMO should not be necessary.

djc commented:
(It might be useful to start split some of the commits out of this PR, since it's obviously complex/contentious -- would be useful to start landing some parts to avoid more rebasing. At least the first commit is an obvious candidate.)

djc commented:
I think the correcting faulty output commits are also independent of the other tricky parts?

FranciscoTGouveia commented:
> I think the correcting faulty output commits are also independent of the other tricky parts?

Yes, they are. Would it be okay if I opened a PR for that now to help streamline this one?

djc commented:
> Yes, they are. Would it be okay if I opened a PR for that now to help streamline this one?

Definitely!

FranciscoTGouveia committed:
feat(downloads): remove the mpsc channel in favor of a `FuturesUnordered`

As installations are still done sequentially, we do not need to have a
mpsc channel and can instead resort to a `FuturesUnordered`.

rami3l commented:
Closing as superseded by #4605. I'd still like to express my great appreciation for you and for all the work you were able to do in the concurrent rustup GSoC and even more. It really means a lot for us to prove what is possible with this project, and I'm glad that we have been able to push the project through the finish line together :)
